use wallet-adapter lib

jordansexton opened the pull request:
This PR replaces the wallet context in oyster/common with modules from the wallet-adapter library and refactors the other packages accordingly. Use of `wallet: any` has been replaced with a type compatible with the context. Tested manually using oyster/lending with Phantom, Torus, and Ledger on Chrome, and SolflareX and Torus on Firefox.

jordansexton commented on packages/bridge/src/contexts/ethereum.tsx:
```diff
@@ -168,6 +171,7 @@ export const EthereumProvider: FunctionComponent = ({ children }) => {
       });
       // @ts-ignore
       wallet.on('accountsChanged', accounts => {
+        // @ts-ignore
```
These `@ts-ignore` comments were added because `accountsChanged` is not a valid event emitted by the adapter interface, and the interface is being used to describe the incompatible WalletConnect and Metamask adapters specific to this package. The reason this compiled previously is because `wallet` was typed as `any` in most places.

jordansexton commented on packages/common/src/components/Settings/index.tsx:
```diff
@@ -31,26 +34,30 @@ export const Settings = ({
             <Button
               type="primary"
-              onClick={disconnect}
+              onClick={() => disconnect().catch()}
```
Asynchronous background errors are handled by the context's global error handler.

jordansexton commented on packages/common/src/contexts/connection.tsx:
```diff
@@ -308,6 +312,7 @@ export const sendTransactions = async (
         successCallback(txid, i);
       })
       .catch(reason => {
+        // @ts-ignore
```
Added because `failCallback(signedTxns[i], i);` fails compilation -- transactions aren't strings.

jordansexton commented on packages/common/src/contexts/wallet.tsx:
```diff
@@ -1,241 +1,160 @@
-    },
-  };
-};
+export type WalletSigner = Pick<WalletAdapter, 'publicKey' | 'signTransaction' | 'signAllTransactions'>;
```
The minimal interface of the context and adapter needed to sign transactions.

SebastianBor replied:
+1 like

jordansexton commented:
<img width="414" alt="Screen Shot 2021-08-12 at 5 53 31 PM" src="https://user-images.githubusercontent.com/1173161/129280095-b7e35d61-25ae-4e5e-9f2e-5784b10e099e.png">

The cancel button is showing up with Chinese text for me. No idea why. Do we even want a cancel button? Here's what it looks like with no footer:

<img width="416" alt="Screen Shot 2021-08-12 at 5 54 37 PM" src="https://user-images.githubusercontent.com/1173161/129280179-b0702ce3-ebd3-4e5d-8e17-54e182b7d409.png">

jordansexton commented:
I'll change the Torus logo to the blue they use. This will fix it not showing up against the black background of the UI.

bartosz-lipinski commented:
@SebastianBor can you try this PR with governance?

jordansexton commented:
Torus and Phantom logos updated

SebastianBor commented on packages/common/src/components/CurrentUserBadge/index.tsx:
```diff
@@ -73,8 +58,8 @@ export const CurrentUserBadge = (props: {
-          {name && <span style={{ marginRight: '0.5rem' }}>{name}</span>}
-          {image}
+          <span style={{ marginRight: '0.5rem' }}>{wallet.name}</span>
+          <img src={wallet.icon} style={iconStyle} />
```
Let's show the wallet address Indenticon and shortened address here. It's more important than the wallet type and name. 
I think there is no need to show the wallet type name at all, but if it looks good we might show the wallet type icon. 

Here is a screenshot from `Governance` design

![image](https://user-images.githubusercontent.com/5817016/129734144-7de52f81-dcad-482e-a77a-13c9e9e916af.png)

jordansexton replied:
I was confused by the original code here, because it seems like if there's a wallet icon, it will use that instead of the Identicon. I may have misunderstood the intent, but I think showing the wallet icon and address will be good. I'll rely on you to tell me how useful the Identicon is, since it doesn't seem that helpful to me.

SebastianBor replied:
Ideally we should have a way to add an avatar to a wallet address and display it here, but without it I would think the Identicon is the next best thing we can do. I personally recognise my wallets using it, but it might be because they are just for different wallets so the wallet type icon would work as well. I don't really have a strong opinion here so we can try to use the wallet type icon instead

jordansexton replied:
Fixed by 06089d1e4311dd960e51752affbfb889b4a6e34b. Uses the wallet icon with the abbreviated address.

SebastianBor commented on packages/common/src/contexts/accounts.tsx:
```diff
@@ -314,23 +314,21 @@ export const getCachedAccount = (
       }
     },
-    [wallet],
+    [publicKey],
```
This dependency won't work if the key instance changes for the same key. The keys need to be compared using string representation `.toBase58()`.
I know it was like that before and this is not related to this change but wanted to point it out.

jordansexton replied:
I don't understand -- the `PublicKey` instance is created when the `WalletAdapter` is connected, and passed through the React context. It should never change unless the wallet reconnects or changes accounts (not implemented yet), and so it should be safe to memoize against.

SebastianBor replied:
If the instance stays the same then it's ok, but if it does for example when reconnecting do we want to revaluate it?

jordansexton replied:
If the adapter disconnects, it will first be set to `null` (in the adapter and in the context) before being set to a `PublicKey`, so on reconnecting even if it was the same instance (which I don't think it can be currently because all the adapters do `new PublicKey(bytes)` on connecting), it should trigger this to be updated anyway since the last value was `null`.

SebastianBor commented on packages/common/src/contexts/wallet.tsx:
```diff
@@ -1,241 +1,160 @@
-    () => WALLET_PROVIDERS.find(({ url }) => url === providerUrl),
-    [providerUrl],
-  );
+export const WalletModal: FC = () => {
```
nit: We don't quite have consistency in the repo :) but I would prefer to just use function here and avoid explicit type specification.

jordansexton replied:
Your call, though I think the type is helpful in case a component fails to return a valid `ReactNode` or `null` value. What's the reason to not have them?

SebastianBor replied:
Mainly personal preferences. In general I prefer when types are inferred rather than being specified explicitly. There are also small technical nuances. For example the FC component would have children prop even though it doesn't use them. 

Regarding the result check without the explicit type if it returned an invalid value then you would know as soon as you would try to use it, so I think it's not that important here.

jordansexton replied:
Yeah, makes sense. I'll remove here then.

SebastianBor commented on packages/common/src/wallet-adapters/solong/index.tsx:
```diff
@@ -1,70 +0,0 @@
-
-    // Temp. workaround to ensure requests to sign multiple transactions at once don't fail when Solong wallet is used
-    // Signing transactions one by one as a fallback works but the UX is not great because user is asked to sign multiple times
-    for (let t of transactions) {
```
This was not the best UX but it did work as a fallback. After the change trying to sign multiple transactions with solong throws error. 
You can reproduce by starting governance [devtools](http://localhost:3000/#/devtools) and trying to generate artefacts with Solong wallet

jordansexton replied:
Great, thanks for finding this. I'll sort it out.

jordansexton replied:
This should be resolved by this change: https://github.com/solana-labs/wallet-adapter/blame/master/packages/wallets/solong/src/adapter.ts#L108 in https://github.com/solana-labs/wallet-adapter/commit/183988deca18478c2d1408a3106288c00cfe5deb

The issue is that the function was expecting Solong to return the transaction object, and it returns undefined. This is fixed by all wallets returning the signed transaction object if the wallet returns nothing.

SebastianBor reviewed (commented):
I like how clean the wallet interface is now and love it reconnects after refreshing.

I've done some smoke testing for governance and it looked ok. Please sync with latest master so I can test the latest version (there were some breaking changes there)

SebastianBor commented:
> <img alt="Screen Shot 2021-08-12 at 5 53 31 PM" width="414" src="https://user-images.githubusercontent.com/1173161/129280095-b7e35d61-25ae-4e5e-9f2e-5784b10e099e.png">
> 
> The cancel button is showing up with Chinese text for me. No idea why. Do we even want a cancel button? Here's what it looks like with no footer:
> 
> <img alt="Screen Shot 2021-08-12 at 5 54 37 PM" width="416" src="https://user-images.githubusercontent.com/1173161/129280179-b0702ce3-ebd3-4e5d-8e17-54e182b7d409.png">

I don't think we need the footer

jordansexton commented on packages/bridge/src/components/CurrentUserWalletBadge/index.tsx:
```diff
@@ -1,12 +1,12 @@
 
 export const CurrentUserWalletBadge = (props: { showDisconnect?: boolean }) => {
-  const { wallet, disconnect } = useWallet();
+  const { wallet, publicKey, disconnect } = useWallet();
```
Hmm. I would hope not to change this in the adapter interface itself or in the `react` package because `wallet.publicKey` is used in a bunch of different codebases in the wild, and this is easy to port over to right now with
```
const wallet = useWallet(); // can use wallet.publicKey now
```

What about just aliasing it here and where needed as `const { publicKey: walletPublicKey } = useWallet();`?

SebastianBor replied:
If it's difficult to change in the api then let's leave it as is. My guess is  most people won't alias it so we most likely  would only add more inconsistency.

jordansexton commented:
Looks like there are some merge conflicts, resolving now.

jordansexton commented:
1. Resolved conficts.
2. Added a missing dev dependency (was causing an issue linting in WebStorm).
3. In the governance package, replaced use if `IWallet` interface with `WalletSigner` from @oyster/common which is used everywhere now.
